docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,289 @@
+# Automatic reachable services
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/6551
+
+## Context and Problem Statement
+
+With large enough Mesh (more than 1000 services) processing the whole cluster state for each proxy is not possible due to performance reasons. 
+The complexity of this computation is O(n^2) (n is the number of services), it also affects Envoy configs size (reaching multiple MB).
+This is why a feature called [reachable services](https://github.com/kumahq/kuma-website/blob/045ff9868acdb4b322759dde3f8053f41948100f/app/_src/production/upgrades-tuning/fine-tuning.md#L8-L7) was introduced.
+Reachable services is a step in the right direction but defining the whole service graph manually by the service owners or the platform team is a big ask (it can be another reason to resist adopting the mesh).
+
+How can we make Kuma easy to use and not have people needing to do systematic performance tuning?
+
+## Glossary
+
+- MTP - [MeshTrafficPermission](https://kuma.io/docs/2.4.x/policies/meshtrafficpermission/#meshtrafficpermission-(beta))
+- ODCDS - [On-demand CDS Updates](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/on_demand_updates_filter#config-http-filters-on-demand)
+
+## Decision Drivers 
+
+* feasibility of implementation for 2.5 release
+* minimize the manual labour needed
+* smooth migration to Mesh for large systems without downtime or service communication issues
+
+## Considered Options
+
+* Using MeshTrafficPermission to generate reverse graph that would feed into reachable services
+* Dynamically load clusters using [ODCDS](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/on_demand/v3/on_demand.proto)
+* Using MeshHTTPRoute / MeshTCPRoute to populate the clusters
+
+## Decision Outcome
+
+Chosen option: "Using MeshTrafficPermission to generate reverse graph that would feed into reachable services"
+
+### Positive Consequences
+
+* not invasive, plug and play, can be removed if proven not efficient enough
```

Comment:
How does it work if you have both old reachable services and new MTP reachable services?

Reply:
they are combined, as mentioned in some other section you can have coarse grain MTP like from namespace to namespace and selected services in that namespace can have specific `reachable-services` will add this to the list

Reply:
They are not combined. Static reachable services takes precedence

Reply:
So if owner of service `b` defines reachable service to `c` and `c` says only `a` and `job` can reach them then service `b` will get the cluster (know the IP) but will bounce back on RBAC. @lahabana what do you think about that?

Reply:
I'm a little concerned with this if we extend the reachable service expressiveness (For .e.g `reachableService: {prefix: foo}`). As MTP would be a way to further reduce the number of services.
Also @slonka 's point makes a little sense from a security POV.

What I first though what we would do is something like:

```go
var allServices []services
if (hasReachableServices) {
   allServices = reachableServices
} else {
   allServices = getAllServices()
}
allServices = filterByPermissionGraph(allServices)
``` 

What's the motivation for replacing instead of combining?

Reply:
For me combined = services from static rs and services from graph.

We have this implem above in the current PR. We could do this, but I feel it's better to have an "escape path" if anything goes wrong.

Reply:
We talked on zoom and agreed that reachable services should take precedence because:
- we can change it back if we need to
- we have escape path if there is an edge case that would incorrectly generate the graph